2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+
+An opportunistic implementation will support an exclusion value in the experimental feature declaration
+for cases where an author is intentionally providing alternative/incompatible implementations.
+
+### Static Code Analysis
+
+Static code analysis support should be provided to address experimental feature authoring.
+Two areas are particularly important:
+
+1. Validating whether an experimental feature is marked correctly.
+   A use case for this occurs when an experimental feature is referenced in a module but not, or incorrectly defined in the manifest.
+   From a script perspective, a `PSScriptAnalyzer` rule would be a reasonable solution.
+   From a compiled code perspective, a reflection-based tool may be needed.
+
+2. Validating whether references to an non-existing experimental feature have been removed.
+   At some point, the experiment is completed and either removed from or incorporated into the code base.
+   At this point, a tool to scan script or compiled code would be needed to ensure any references to the experiment have been removed.
+
+## Open Issues
+
+### Whether experimental features can be included in PowerShell GA releases
+
+It's arguable whether or not experimental features should be allowed in PowerShell GA releases.
+Take C# an an instance, it has experimental language features while in the development of a new version.
+However, when reaching the final release,
+there won't be any experimental features -- a feature is either officially kept or removed from the final release.
+
+The benefit of not allowing experimental features in a GA release is to improve stability.
+But the drawback is that we cannot get more usage and feedback for an experimental feature
+before deciding to keep or cut the feature.
+This is because preview releases of PowerShell usually get much less usage compared with the GA release.
+
+### Whether experimental feature dependencies should be parsed when reading the configuration
+
+When reading the experimental feature list from configuration file,
+shall we parse the dependencies at the same time or defer the dependency resolution until module loading?
+
+As described in the [`Experimental Feature Dependencies`](#experimental-feature-dependencies) section,
+an experimental feature may have a chain dependency on other experimental features.
+In that section, the proposal is to check for dependencies at module loading time.
+
+Another proposal is to resolve the dependencies when loading the configuration file,
+so that PowerShell can fail early when dependencies are not satisfied.
+The fail-early idea is desired,
+however, PowerShell loads the enabled experimental feature list at very early stage,
+where the engine is not ready for the operations required for resolving dependencies,
+such as module discovery.
+Also, there will be an impact to the startup time if we do this resolution when loading the configuration file.
```

Comment:
Since Experimental features are non-GA it seems like a modest start-up impact is not too bad.  With the extra module processing it seems likely that command discovery will be slower in general too.  I think we would want to ensure the module analysis cache file is updated to include experimental information.

Reply:
Slowdown startup is not acceptable for GA.

Reply:
That is what I said.  So for non-GA a modest impact should be Ok.